libtenzir/builtins/operators/from_opensearch.cpp

```diff
@@ -0,0 +1,480 @@
+                        }
+                        for (const auto& chunk : split_at_newline(ptr)) {
+                          if (chunk.empty()) {
+                            continue;
+                          }
+                          auto view = simdjson::padded_string_view{
+                            reinterpret_cast<const char*>(chunk.data()),
+                            chunk.size(),
+                            chunk.capacity(),
+                          };
+                          parser.parse(view);
+                        }
+                        if (parser.abort_requested) {
+                          return slices;
+                        }
+                        parser.validate_completion();
+                        if (parser.abort_requested) {
+                          return slices;
+                        }
+                        // Get all remaining events
+                        auto result = parser.builder.finalize_as_table_slice();
+                        if (keep_actions) {
+                          for (auto&& slice : result) {
+                            slices.push_back(std::move(slice));
+                          }
+                          return slices;
+                        }
+                        auto is_action = true;
+                        for (auto&& slice : result) {
+                          slices.push_back(
+                            handle_slice(is_action, std::move(slice)));
+                        }
+                        return slices;
+                      });
+                  })
+                  .to_typed_stream<slice_result>("from_opensearch",
+                                                 std::chrono::seconds{1}, 1);
+          });
+    if (not server) {
+      diagnostic::error("failed to setup http server: {}", server.error())
+        .primary(args_.op)
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+    auto _guard
+      = detail::scope_guard{[server = std::move(*server)]() mutable noexcept {
+          // NOTE: This could very well throw but scope_guard needs to be
+          // noexcept.
+          server.dispose();
+        }};
+    TENZIR_ASSERT(stream);
+    ctrl.self().observe(*stream, 30, 10).for_each([&](slice_result result) {
+      match(
+        std::move(result),
+        [](std::monostate) {
+          TENZIR_UNREACHABLE();
+        },
+        [&](diagnostic&& d) {
+          ctrl.diagnostics().emit(std::move(d));
+        },
+        [&](std::vector<table_slice>&& input) {
+          ctrl.set_waiting(false);
+          std::move(input.begin(), input.end(), std::back_inserter(slices));
+        });
+    });
+    launch();
+    while (true) {
+      ctrl.set_waiting(true);
+      co_yield {};
+      // NOTE: Must be an index-based loop. Johannes found out that the thread
+      // can go back to the observe loop after yielding here, causing the
+      // vector's iterator to be invalidated.
```

Comment:
* It's nice to give credit, but it doesn't really enhance the explanation in a code comment.
* It's a good comment explaining the why, but I'm still confused why index-based access fares better than an iterator here. This only works if the vector can only grow.

Reply:
The vector is only cleared after the loop, nothing else removes elements from the vector.